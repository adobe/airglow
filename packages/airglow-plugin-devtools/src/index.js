/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import React from 'react';
import { POST_ENHANCER, COMPONENT, COMPOSER } from 'airglow';
import Devtools from './devtools';

export default function devtoolsPlugin({
  reduxDevtools = false,
  reduxBrowser = false
} = {}) {
  return (engine) => {
    if (reduxDevtools) {
      engine.plugin(POST_ENHANCER, Devtools.instrument());
      engine.plugin(COMPONENT, <Devtools key="c" />);
    }
    if (reduxBrowser) {
      engine.plugin(COMPOSER, browserCompose);
    }
  };
}

/* eslint-disable no-underscore-dangle */
const browserHook = () => global.window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__;

const browserCompose = compose =>
  (browserHook() ? browserHook()({}) : compose);
