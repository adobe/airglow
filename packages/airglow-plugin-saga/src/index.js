/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import createSagaMiddleware from 'redux-saga';
import { FEATURE, MIDDLEWARE, BOOTSTRAP_MODULE } from 'airglow';

export const SAGA = 'wire_sagas';

export default function sagaPlugin() {
  const middleware = createSagaMiddleware();

  return (engine) => {
    engine.plugin(MIDDLEWARE, middleware);
    engine.plugin(FEATURE, sagaFeature(middleware));
    engine.plugin(BOOTSTRAP_MODULE, bootstrapSagas(middleware));
  };
}

const sagaFeature = middleware => lookup =>
  lookup(SAGA).forEach(
    plugin => middleware.run(plugin)
  );

const bootstrapSagas = middleware => ({ sagas }) => {
  if (!sagas) { return; }
  const runSagas = Array.isArray(sagas) ? sagas : [sagas];
  runSagas.forEach(saga => middleware.run(saga));
};
