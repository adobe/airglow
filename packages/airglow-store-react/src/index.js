/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import * as R from 'ramda';

import { REDUCER, HOC } from 'airglow';
import { StoreProvider } from './store.context';

export { default as connect } from './connect';

let SUBSCRIBE_INDEX = 0;

class Store {
  constructor(reducers = []) {
    this.state = {};
    this.reducers = reducers;
    this.dispatch = this.dispatch.bind(this);
    this.injectReducers = this.injectReducers.bind(this);
    this.subscriptions = {};
    this.dispatch({ type: 'REACT.STORE.INITIALIZE' });
  }

  dispatch(action) {
    this.state = R.reduce(
      (newState, reducer) => reducer(newState, action),
      this.state,
      this.reducers || []
    );
    for (let i = 0; i <= SUBSCRIBE_INDEX; ++i) {
      if (this.subscriptions[i]) {
        this.subscriptions[i](this);
      }
    }
  }

  getState() {
    return this.state;
  }

  subscribe(callback) {
    this.subscriptions[++SUBSCRIBE_INDEX] = callback;
    return () => { this.subscriptions[SUBSCRIBE_INDEX] = null; };
  }

  injectReducers(newReducers) {
    if (!newReducers) { return; }
    this.reducers = R.concat(this.reducers, newReducers);
    this.dispatch({ type: 'REACT.STORE.INJECT.REDUCERS' });
  }
}

export const createStore = reducers => new Store(reducers);

class StoreFactory {
  create(lookup) {
    this.store = createStore(getReducers(lookup));
    return this.store;
  }

  bootstrap(engine) {
    engine.plugin(HOC, {
      component: StoreProvider,
      props: {
        store: this.store
      }
    });
  }
}

const getReducers = lookup => reduceLookup(lookup(REDUCER));

const reduceLookup = data => data.reduce(
  (results, plugin) => mergeInto(results, plugin), []
);

const mergeInto = (source, toMerge) =>
  source.concat(Array.isArray(toMerge) ? toMerge : [toMerge]);

export default () => new StoreFactory();
