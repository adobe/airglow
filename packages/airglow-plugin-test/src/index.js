/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import { REDUCER } from 'airglow';
import { slice } from '@airglow/reducers';
import { last } from 'ramda';
import reducer from './reducer';

const STORE_KEY = '_test_dispatches';

export default function testPlugin() {
  return (engine) => {
    engine.plugin(REDUCER, slice(STORE_KEY).with(reducer));
  };
}

export const lastDispatch = state => last(state[STORE_KEY]);
export const dispatchList = state => state[STORE_KEY];
