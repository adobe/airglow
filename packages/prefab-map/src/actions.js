/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import { createAction } from '@airglow/reducers';

export const TOGGLE_MODE = 'prefab.map.toggle.mode';
export const BUTTON_ZOOM = 'prefab.map.button.zoom';
export const ZOOM = 'prefab.map.zoom';
export const CHANGE_POSITION = 'prefab.map.change.position';

const constructAction = type => construct => createAction(type,
  ({ ...other } = {}) => ({ construct, ...other }));

export const doToggleMode = constructAction(TOGGLE_MODE);
export const doZoom = constructAction(ZOOM);
export const doButtonZoom = (construct, size) => constructAction(BUTTON_ZOOM)(construct)({ size });
export const doChangePosition = constructAction(CHANGE_POSITION);
