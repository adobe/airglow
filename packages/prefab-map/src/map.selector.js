/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import * as R from 'ramda';
import * as A from './actions';
import { DEFAULT_ZOOM } from './constants';

const zoom = name => state =>
  R.defaultTo(
    constructPath(name, 'defaultZoom')(state) || DEFAULT_ZOOM,
    storePath(name, 'zoom')(state)
  );

const constructPath = (name, prop) => R.path(['prefab', name, 'construct', prop]);
const storePath = (name, prop) => R.path(['prefab', name, 'store', prop]);

const mode = name =>
  state => (storePath(name, 'mode')(state) ? 'satellite' : 'map');

const position = name =>
  state => storePath(name, 'position')(state);

export default function mapSelector(name) {
  return {
    position: position(name),
    mode: mode(name),
    zoom: zoom(name),
    toggleModeAction: A.doToggleMode(name),
    positionAction: A.doChangePosition(name),
    zoomOutAction: () => A.doButtonZoom(name, -1),
    zoomInAction: () => A.doButtonZoom(name, 1),
    zoomAction: zoomIn => A.doZoom(name)({ zoom: zoomIn })
  };
}
