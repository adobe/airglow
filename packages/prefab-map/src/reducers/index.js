/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import { call } from '@airglow/reducers';
import * as T from '../actions';
import toggleModeReducer from './toggle.mode.reducer';
import { buttonZoomReducer, zoomReducer } from './zoom.reducer';
import positionReducer from './position.reducer';

export default call(
  call(buttonZoomReducer).for(T.BUTTON_ZOOM),
  call(zoomReducer).for(T.ZOOM),
  call(toggleModeReducer).for(T.TOGGLE_MODE),
  call(positionReducer).for(T.CHANGE_POSITION)
);
