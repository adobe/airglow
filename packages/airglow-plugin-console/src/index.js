/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import * as R from 'ramda';
import { REDUCER, BOOTSTRAP_MODULE, BOOTSTRAP_PLUGIN } from 'airglow';
import { slice, createAction } from '@airglow/reducers';
import { BOOTSTRAP_CONSOLE } from './actions';
import reducer from './reducers';

const STORE_KEY = 'console-commands';
const doBootstrap  = createAction(BOOTSTRAP_CONSOLE);

export default function prefabPlugin(config) {
  return (engine) => {
    engine.plugin(REDUCER, slice(STORE_KEY).with(reducer));
    engine.plugin(BOOTSTRAP_MODULE, bootstrap(engine));
    engine.plugin(BOOTSTRAP_PLUGIN, bootstrapPlugin(engine, config));
  };
}

const logHelp = ({ state }) => {
  R.forEachObjIndexed(
    ({ description }, key) => {
      console.info(`${key}(): ${description}`); // eslint-disable-line no-console
    },
    state['console-commands']
  );
};

const bootstrapPlugin = (
  engine, { help = 'help', helpDescription = 'Display this help menu', consoleCommands }
) => ({ dispatch }) => {
  bootstrap(engine)({
    dispatch,
    consoleCommands: {
      [help]: { description: helpDescription, command: logHelp },
      ...consoleCommands
    }
  });
};

const bootstrap = engine => ({ dispatch, consoleCommands }) => {
  if (consoleCommands) {
    dispatch(doBootstrap(consoleCommands));
    R.forEachObjIndexed(
      ({ command }, key) => {
        window[key] = () => command({ state: engine.getStore().getState(), dispatch });
      },
      consoleCommands
    );
  }
};
