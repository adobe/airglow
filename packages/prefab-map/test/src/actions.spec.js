/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import * as A from '../../src/actions';

it('should create a toggle action', function () {
  expect(A.doToggleMode('myMap')()).toMatchSnapshot();
});

it('should create a zoom action', function () {
  expect(A.doZoom('myMap')({ zoom: 3 })).toMatchSnapshot();
});

it('should create a button zoom action', function () {
  expect(A.doButtonZoom('myMap', { size: -1 })).toMatchSnapshot();
});

it('should create a position action', function () {
  expect(A.doChangePosition('myMap')({ latitide: 3, longitude: -5 })).toMatchSnapshot();
});
