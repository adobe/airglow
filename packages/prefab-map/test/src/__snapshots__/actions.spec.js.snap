// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create a button zoom action 1`] = `
Object {
  "payload": Object {
    "construct": "myMap",
    "size": Object {
      "size": -1,
    },
  },
  "type": "prefab.map.button.zoom",
}
`;

exports[`should create a position action 1`] = `
Object {
  "payload": Object {
    "construct": "myMap",
    "latitide": 3,
    "longitude": -5,
  },
  "type": "prefab.map.change.position",
}
`;

exports[`should create a toggle action 1`] = `
Object {
  "payload": Object {
    "construct": "myMap",
  },
  "type": "prefab.map.toggle.mode",
}
`;

exports[`should create a zoom action 1`] = `
Object {
  "payload": Object {
    "construct": "myMap",
    "zoom": 3,
  },
  "type": "prefab.map.zoom",
}
`;
